File: backend/internal/utils/timezone.go
package utils

import (
	"strings"
	"time"
)

// StateTimezones maps US state (and DC) abbreviations to IANA timezone names.
//
// This is the full 50-state + DC map. It MUST stay in sync with the writer-side
// maps the show data was actually stored under — cli/src/lib/timezone.ts and
// frontend/lib/utils/timeUtils.ts (PSY-1009) — because the venue-timezone
// backfill (PSY-987) derives the zone a date-only show was *written* under from
// this map: a short map (defaulting most states to Phoenix) would make a
// correctly-stored explicit-time show in an unmapped state read as a false 20:00
// Phoenix default and be wrongly re-anchored. States that span two zones use
// their predominant zone (the same approximation the writers made).
//
// The three maps are identical as of PSY-987, but the sync is NOT enforced by
// CI (they're Go vs TypeScript): TestGetTimezoneForState_FullMapCoverage only
// guards this Go map's own coverage. If you edit any of the three, edit all
// three — drift here silently re-opens the re-anchor corruption class.
var StateTimezones = map[string]string{
	// Pacific
	"CA": "America/Los_Angeles",
	"NV": "America/Los_Angeles",
	"OR": "America/Los_Angeles",
	"WA": "America/Los_Angeles",
	// Mountain
	"AZ": "America/Phoenix", // no DST
	"CO": "America/Denver",
	"NM": "America/Denver",
	"MT": "America/Denver",
	"UT": "America/Denver",
	"WY": "America/Denver",
	"ID": "America/Boise",
	// Central
	"IL": "America/Chicago",
	"TX": "America/Chicago",
	"AL": "America/Chicago",
	"AR": "America/Chicago",
	"IA": "America/Chicago",
	"KS": "America/Chicago",
	"LA": "America/Chicago",
	"MN": "America/Chicago",
	"MO": "America/Chicago",
	"MS": "America/Chicago",
	"ND": "America/Chicago",
	"NE": "America/Chicago",
	"OK": "America/Chicago",
	"SD": "America/Chicago",
	"TN": "America/Chicago",
	"WI": "America/Chicago",
	"IN": "America/Indiana/Indianapolis",
	// Eastern
	"NY": "America/New_York",
	"CT": "America/New_York",
	"DC": "America/New_York",
	"DE": "America/New_York",
	"FL": "America/New_York",
	"GA": "America/New_York",
	"KY": "America/New_York",
	"MA": "America/New_York",
	"MD": "America/New_York",
	"ME": "America/New_York",
	"MI": "America/New_York",
	"NC": "America/New_York",
	"NH": "America/New_York",
	"NJ": "America/New_York",
	"OH": "America/New_York",
	"PA": "America/New_York",
	"RI": "America/New_York",
	"SC": "America/New_York",
	"VA": "America/New_York",
	"VT": "America/New_York",
	"WV": "America/New_York",
	// Non-contiguous
	"AK": "America/Anchorage",
	"HI": "Pacific/Honolulu",
}

// DateOnlyEventHour is the venue-local hour a show carrying a DATE but no time
// of day is anchored at.
//
// The alternative is storing bare UTC midnight, and that is the bug this
// constant exists to keep closed. shows.event_date is a TIMESTAMPTZ with no
// companion precision flag, so once a date-only value is written as 00:00:00Z
// nothing downstream can recover the fact that its time was never known. In
// every US zone that instant is the PREVIOUS evening local, so the show sorts
// and renders under the wrong calendar day, and any "still upcoming" bound drops
// it from the moment UTC midnight passes — 17:00 the day before in Phoenix.
// PSY-1780, PSY-1849 and PSY-1861 are three separate bites of that one wound.
//
// 20:00 is not a guess at any particular door time; it is a convention chosen so
// that the anchored instant lands unambiguously on the intended calendar EVENING
// in every US zone, which is the only property readers depend on.
//
// Exported and referenced by every writer rather than restated as a literal:
// discovery imports (services/pipeline.parseEventDate) and community request
// fulfillment (handlers/community.parseShowEventDate) must agree, or a reader
// inspecting a row cannot tell which writer produced it. The frontend show form
// (frontend/features/shows/components/show-form-utils.DEFAULT_EVENT_TIME, which
// a blank time field falls back to) and the ph CLI are writers too and are NOT
// compile-checked against this — keep them in sync by hand, exactly as
// StateTimezones warns above.
//
// NOT enforced on the show create/update API, and deliberately so. Those
// endpoints take event_date as an already-parsed instant, and an instant cannot
// carry the fact that its time was never known: the frontend form serializes
// 20:00 venue-local to a plain UTC string, so a genuine 8pm show at an Eastern
// venue in summer arrives as exactly 00:00:00Z and is indistinguishable from a
// bare date. Closing that path needs an explicit precision signal in the
// contract (a YYYY-MM-DD variant, the way parseShowEventDate already accepts
// one), not a rule inferred from the instant. See PSY-1861.
const DateOnlyEventHour = 20

// GetTimezoneForState returns the IANA timezone for a US state abbreviation.
// Defaults to "America/Phoenix" if the state is not found.
func GetTimezoneForState(state string) string {
	if tz, ok := StateTimezones[strings.ToUpper(state)]; ok {
		return tz
	}
	return "America/Phoenix"
}

// HasTimezoneForState reports whether StateTimezones carries this state, i.e.
// whether GetTimezoneForState answers from the map rather than from its
// America/Phoenix default.
//
// The distinction the default erases: "AZ" and "England" both come back as
// America/Phoenix, and only the first of them is an answer. A caller about to
// print a WALL CLOCK needs to tell them apart, because the second is wrong by
// hours; a caller printing a calendar DATE does not, because the fallback day is
// still the best available one.
//
// Case folding matches GetTimezoneForState exactly, and neither trims, so the
// two always agree on which inputs the map holds.
func HasTimezoneForState(state string) bool {
	_, ok := StateTimezones[strings.ToUpper(state)]
	return ok
}

// EventLocation resolves the IANA location for rendering an event time in a
// venue's local zone. Precedence: a valid explicit venue timezone, then the US
// state->tz map (GetTimezoneForState, which itself defaults unknown/empty input
// to America/Phoenix), and finally UTC only if a non-empty timezone string
// fails to load. A malformed venue timezone falls through to the state map
// rather than jumping straight to UTC. (PSY-996)
func EventLocation(timezone *string, stateFallback string) *time.Location {
	if timezone != nil && *timezone != "" {
		if loc, err := time.LoadLocation(*timezone); err == nil {
			return loc
		}
		// Malformed/unknown IANA string — fall through to the state map below.
	}
	if loc, err := time.LoadLocation(GetTimezoneForState(stateFallback)); err == nil {
		return loc
	}
	return time.UTC
}

File: cli/src/lib/timezone.ts
/**
 * Timezone utilities for converting local venue times to UTC.
 *
 * The state to timezone mapping mirrors the frontend's getTimezoneForState()
 * in frontend/lib/utils/timeUtils.ts. Keep them in sync.
 *
 * `localTimeToUTC` and the frontend's `combineDateTimeToUTC` no longer agree:
 * this one probes the zone offset twice and matches Go for a wall clock inside a
 * DST transition, the frontend's probes once and does not. Reconciling them is a
 * frontend change with its own gates, not something to do quietly from here.
 */

/** Map of US state abbreviations to IANA timezones. */
const STATE_TIMEZONES: Record<string, string> = {
  AZ: "America/Phoenix",
  CA: "America/Los_Angeles",
  NV: "America/Los_Angeles",
  CO: "America/Denver",
  NM: "America/Denver",
  TX: "America/Chicago",
  NY: "America/New_York",
  // Eastern
  CT: "America/New_York",
  DC: "America/New_York",
  DE: "America/New_York",
  FL: "America/New_York",
  GA: "America/New_York",
  MA: "America/New_York",
  MD: "America/New_York",
  ME: "America/New_York",
  MI: "America/New_York",
  NC: "America/New_York",
  NH: "America/New_York",
  NJ: "America/New_York",
  OH: "America/New_York",
  PA: "America/New_York",
  RI: "America/New_York",
  SC: "America/New_York",
  VA: "America/New_York",
  VT: "America/New_York",
  WV: "America/New_York",
  // Central
  AL: "America/Chicago",
  AR: "America/Chicago",
  IA: "America/Chicago",
  IL: "America/Chicago",
  IN: "America/Indiana/Indianapolis",
  KS: "America/Chicago",
  KY: "America/New_York",
  LA: "America/Chicago",
  MN: "America/Chicago",
  MO: "America/Chicago",
  MS: "America/Chicago",
  ND: "America/Chicago",
  NE: "America/Chicago",
  OK: "America/Chicago",
  SD: "America/Chicago",
  TN: "America/Chicago",
  WI: "America/Chicago",
  // Mountain
  ID: "America/Boise",
  MT: "America/Denver",
  UT: "America/Denver",
  WY: "America/Denver",
  // Pacific
  OR: "America/Los_Angeles",
  WA: "America/Los_Angeles",
  // Non-contiguous
  AK: "America/Anchorage",
  HI: "Pacific/Honolulu",
};

/**
 * Get IANA timezone for a US state abbreviation.
 * Defaults to America/Phoenix (Arizona, no DST) — same as frontend.
 */
export function getTimezoneForState(state: string): string {
  return STATE_TIMEZONES[state.toUpperCase()] || "America/Phoenix";
}

/**
 * Whether the state map actually covers this state, as opposed to handing back
 * its America/Phoenix default.
 *
 * The default is indistinguishable from a real answer once getTimezoneForState
 * has returned, so a caller that needs to know whether a show's zone is KNOWN
 * rather than assumed has to ask before resolving. Mirrors hasTimezoneForState
 * in frontend/lib/utils/timeUtils.ts, which the read surfaces gate their clocks
 * on.
 */
export function hasTimezoneForState(state?: string): boolean {
  return !!state && state.toUpperCase() in STATE_TIMEZONES;
}

/**
 * Whether a zone can be resolved for a show from the venue's own zone or from
 * the state map, rather than from the America/Phoenix default.
 *
 * Mirrors isShowTimezoneResolved in frontend/lib/utils/formatters.ts. The read
 * surfaces that print a clock (the status stripe's DOORS / MUSIC / start-time
 * segments, MusicEvent.startDate) all refuse when this is false, so a write path
 * that stores a time on the default stores an instant nothing renders.
 */
export function isShowTimezoneResolved(
  state?: string,
  timezone?: string,
): boolean {
  return (!!timezone && isValidTimeZone(timezone)) || hasTimezoneForState(state);
}

/**
 * Resolve the IANA zone a show's wall-clock time should be read in.
 *
 * Precedence, and the reason it is stated in one place: a venue that is already
 * in the database carries the zone the geocoder resolved for it, and that is the
 * SAME value every read surface renders the show in. The state map is a guess
 * for venues without one, and outside the US it is not even a guess: it hands
 * back America/Phoenix for "England", "QC", or "" alike.
 *
 * PSY-1873: `ph submit-show` keyed on the state map alone, so a date-only show
 * at a venue outside the US was written as 20:00 Phoenix while the show page,
 * which reads venues.timezone, rendered it in the venue's real zone. That is
 * wrong by the offset between the two, which for European venues moves the
 * show onto the following calendar day at around 4 AM. Nearly every show at a
 * non-US venue in production carried that signature.
 *
 * Mirrors the LOGIC of resolveShowTimezone in
 * frontend/lib/utils/formatters.ts, though not its memoization: that one sits
 * on a render path that asks the same question hundreds of times, while this
 * resolves once per show in a batch. An unloadable zone string falls through to
 * the state map rather than crashing, because Intl.DateTimeFormat throws a
 * RangeError on a bad zone.
 */
export function resolveVenueTimezone(
  state?: string,
  timezone?: string,
): string {
  if (timezone && isValidTimeZone(timezone)) return timezone;
  return getTimezoneForState(state || "");
}

/** Whether an IANA zone name exists in this runtime's tz database. */
function isValidTimeZone(name: string): boolean {
  try {
    new Intl.DateTimeFormat("en-US", { timeZone: name });
    return true;
  } catch {
    return false;
  }
}

/**
 * Convert a local date+time in a given timezone to a UTC ISO 8601 string.
 *
 * Offset probing, because JS has no "this wall clock, in this zone" constructor.
 * The probe runs TWICE: the first offset is read at the wall clock interpreted
 * as UTC, which is up to a day away from the answer and lands on the wrong side
 * of any DST transition in between. Re-reading the offset at the candidate
 * instant and re-deriving from it is what makes a clock inside a transition
 * window come out right, and a listing that states 12:30 AM and 1:30 AM on a
 * spring-forward night is exactly such a clock.
 *
 * A wall clock that does not exist (the hour a spring-forward skips) has no
 * correct answer, and one that happens twice (fall-back) has two. For both,
 * this returns the instant Go's `time.Date` returns for the same wall clock and
 * zone, verified against Go across every 2025-2027 transition in 30 zones. That
 * agreement is the whole specification: which side of the transition it lands
 * on varies by zone and is not a rule worth restating here. A caller that must
 * not store a clock the source could not have meant should ask
 * `localClockExists` first.
 *
 * @param dateStr  Date in YYYY-MM-DD format
 * @param timeStr  Time in HH:MM or HH:MM:SS format
 * @param timezone IANA timezone (e.g., "America/Phoenix")
 * @returns ISO 8601 UTC string like "2026-04-15T03:00:00Z"
 */
export function localTimeToUTC(
  dateStr: string,
  timeStr: string,
  timezone: string,
): string {
  const [year, month, day] = dateStr.split("-").map(Number);
  const timeParts = timeStr.split(":").map(Number);
  const hours = timeParts[0];
  const minutes = timeParts[1] || 0;

  // The requested wall clock, read as if it were UTC.
  const wanted = Date.UTC(year, month - 1, day, hours, minutes, 0, 0);

  const formatter = new Intl.DateTimeFormat("en-US", {
    timeZone: timezone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hour12: false,
  });

  /** The wall clock this instant reads as in the zone, as a UTC-shaped number. */
  const wallClockAt = (instant: number): number => {
    const parts = formatter.formatToParts(new Date(instant));
    const p = (type: string) =>
      Number(parts.find((x) => x.type === type)?.value ?? 0);
    let hour = p("hour");
    if (hour === 24) hour = 0; // Intl may return 24 for midnight
    return Date.UTC(p("year"), p("month") - 1, p("day"), hour, p("minute"), 0, 0);
  };

  // The offset read at an instant, and the candidate that offset implies.
  const offsetAt = (instant: number): number => wallClockAt(instant) - instant;
  const first = wanted - offsetAt(wanted);
  const second = wanted - offsetAt(first);

  // `second` is the answer whenever it reads back as the clock that was asked
  // for; `first` covers the case where the re-probe overshot back across the
  // same transition. When neither reads back, the clock does not exist and
  // `second` is the post-transition instant.
  const chosen =
    wallClockAt(second) === wanted
      ? second
      : wallClockAt(first) === wanted
        ? first
        : second;

  // Return as RFC3339 without milliseconds (Go's time.Time expects this)
  return new Date(chosen).toISOString().replace(/\.\d{3}Z$/, "Z");
}

/**
 * Whether this wall clock exists on this day in this zone.
 *
 * False only inside the hour a spring-forward skips, where `localTimeToUTC` must
 * still return something and returns the instant Go does. A writer that refuses
 * to store a time the source did not state asks this first: 2:30 AM on a
 * spring-forward night is not a time that happened, and storing it as 1:30 AM
 * publishes a clock nobody printed.
 */
export function localClockExists(
  dateStr: string,
  timeStr: string,
  timezone: string,
): boolean {
  const instant = Date.parse(localTimeToUTC(dateStr, timeStr, timezone));
  const parts = new Intl.DateTimeFormat("en-GB", {
    timeZone: timezone,
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
  }).formatToParts(new Date(instant));
  const p = (type: string) => parts.find((x) => x.type === type)?.value ?? "";
  const hour = p("hour") === "24" ? "00" : p("hour");
  const [wantedHour, wantedMinute] = timeStr.split(":");
  return hour === wantedHour && p("minute") === wantedMinute;
}
